src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp

```diff
@@ -0,0 +1,213 @@
+                    /// table, then we will fill /metadata_version node in zookeeper.
+                    /// Otherwise on the next restart we can again use version from
+                    /// shared metadata node because it was not changed.
+                    Coordination::Stat metadata_stat;
+                    zookeeper->get(zookeeper_path + "/metadata", &metadata_stat);
+                    storage.metadata_version = metadata_stat.version;
+                }
+            });
+
+        /// Temporary directories contain uninitialized results of Merges or Fetches (after forced restart),
+        /// don't allow to reinitialize them, delete each of them immediately.
+        storage.clearOldTemporaryDirectories(0, {"tmp_", "delete_tmp_", "tmp-fetch_"});
+        storage.clearOldWriteAheadLogs();
+        if (storage.getSettings()->merge_tree_enable_clear_old_broken_detached)
+            storage.clearOldBrokenPartsFromDetachedDirecory();
+
+        withRetries([&] { storage.createNewZooKeeperNodes(zookeeper); });
+        withRetries([&] { storage.syncPinnedPartUUIDs(zookeeper); });
+
+        withRetries([&] { storage.createTableSharedID(zookeeper); });
+
+        finalizeInitialization();
+        notifyIfFirstTry();
+    }
+    catch (const Exception & e)
+    {
+        if (e.code() == ErrorCodes::ABORTED && shutdown_called)
+        {
+            LOG_WARNING(log, "Shutdown called, cancelling initialization");
+            return;
+        }
+
+        LOG_ERROR(log, "Initialization failed, table will remain readonly. Error: {}", e.message());
+
+        {
+            std::lock_guard lock(storage.initialization_mutex);
+            storage.init_phase = StorageReplicatedMergeTree::InitializationPhase::INITIALIZATION_DONE;
+        }
+
+        notifyIfFirstTry();
+    }
+};
+
+void ReplicatedMergeTreeAttachThread::finalizeInitialization()
+{
+
+    LOG_INFO(log, "Table is initialized");
+    using enum StorageReplicatedMergeTree::InitializationPhase;
+    {
+        std::lock_guard lock(storage.initialization_mutex);
+        if (!storage.startup_called)
+        {
+            storage.init_phase = INITIALIZATION_DONE;
+            return;
+        }
+
+        storage.init_phase = STARTUP_IN_PROGRESS;
+    }
+
+    storage.startupImpl(std::move(zookeeper));
+}
+
+void ReplicatedMergeTreeAttachThread::notifyIfFirstTry()
+{
+    if (!first_try_done.exchange(true))
+        first_try_done.notify_one();
+}
+
+void ReplicatedMergeTreeAttachThread::tryReconnect()
+{
+    zookeeper = nullptr;
+    while (!shutdown_called)
+    {
+        try
+        {
+            const auto context = storage.getContext();
+            if (storage.zookeeper_name == storage.default_zookeeper_name)
+                zookeeper = std::make_shared<zkutil::ZooKeeper>(context->getConfigRef(), "zookeeper", context->getZooKeeperLog());
+            else
+                zookeeper = context->getAuxiliaryZooKeeperConnection(storage.zookeeper_name);
```

Comment:
During tests I saw that CH didn't start in time. The problem is that we get a mutex for ZK, and try to connect. That mutex is held for the entire connection timeout (if it can't connect). 
By adding retries every 10 seconds in background, that mutex was a bigger congestion point because we have an extra thread taking the mutex, waiting the connection timeout, and blocking other threads used for creating replicated tables.
With 60 seconds retry period it'll probably work better. I'll try and I can always increase the wait_time in tests.
I still don't like it that we can start a bunch of tables with a nonworking ZK connection forcing for each of them to timeout sequentially.